crates/bevy_ecs/src/world/filtered_resource.rs

```diff
@@ -0,0 +1,576 @@
+/// # struct A;
+/// #
+/// # #[derive(Default, Resource)]
+/// # struct B;
+/// #
+/// # #[derive(Default, Resource)]
+/// # struct C;
+/// #
+/// # let mut world = World::new();
+/// // Use `FilteredResourcesParamBuilder` to declare access to resources.
+/// let system = (FilteredResourcesParamBuilder::new(|builder| {
+///     builder.add_read::<B>().add_read::<C>();
+/// }),)
+///     .build_state(&mut world)
+///     .build_system(resource_system);
+///
+/// world.init_resource::<A>();
+/// world.init_resource::<C>();
+///
+/// fn resource_system(res: FilteredResources) {
+///     // The resource exists, but we have no access, so we can't read it.
+///     assert!(res.get::<A>().is_none());
+///     // The resource doesn't exist, so we can't read it.
+///     assert!(res.get::<B>().is_none());
+///     // The resource exists and we have access, so we can read it.
+///     let c = res.get::<C>().unwrap();
+///     // The type parameter can be left out if it can be determined from use.
+///     let c: Res<C> = res.get().unwrap();
+/// }
+/// #
+/// # world.run_system_once(system);
+/// ```
+///
+/// This can be used alongside ordinary [`Res`] and [`ResMut`](crate::system::ResMut) parameters if they do not conflict.
+///
+/// ```
+/// # use bevy_ecs::{prelude::*, system::*};
+/// #
+/// # #[derive(Default, Resource)]
+/// # struct A;
+/// #
+/// # #[derive(Default, Resource)]
+/// # struct B;
+/// #
+/// # let mut world = World::new();
+/// # world.init_resource::<A>();
+/// # world.init_resource::<B>();
+/// #
+/// let system = (
+///     FilteredResourcesParamBuilder::new(|builder| {
+///         builder.add_read::<A>();
+///     }),
+///     ParamBuilder,
+///     ParamBuilder,
+/// )
+///     .build_state(&mut world)
+///     .build_system(resource_system);
+///
+/// // Read access to A does not conflict with read access to A or write access to B.
+/// fn resource_system(filtered: FilteredResources, res_a: Res<A>, res_mut_b: ResMut<B>) {
+///     let res_a_2: Res<A> = filtered.get::<A>().unwrap();
+/// }
+/// #
+/// # world.run_system_once(system);
+/// ```
+///
+/// But it will conflict if it tries to read the same resource that another parameter writes.
+///
+/// ```should_panic
+/// # use bevy_ecs::{prelude::*, system::*};
+/// #
+/// # #[derive(Default, Resource)]
+/// # struct A;
+/// #
+/// # let mut world = World::new();
+/// # world.init_resource::<A>();
+/// #
+/// let system = (
+///     FilteredResourcesParamBuilder::new(|builder| {
+///         builder.add_read::<A>();
+///     }),
+///     ParamBuilder,
+/// )
+///     .build_state(&mut world)
+///     .build_system(invalid_resource_system);
+///
+/// // Read access to A conflicts with write access to A.
+/// fn invalid_resource_system(filtered: FilteredResources, res_mut_a: ResMut<A>) { }
+/// #
+/// # world.run_system_once(system);
+/// ```
+#[derive(Clone)]
+pub struct FilteredResources<'w> {
+    world: UnsafeWorldCell<'w>,
+    access: Access<ComponentId>,
```

Comment:
Is there a reason this isn't a `&'s Access<ComponentId>`? This way you would avoid having to clone this `Access` in `get_param`.

Reply:
I couldn't make that work with the `From<&World>` impls, since they don't have anywhere to borrow from.  And I want those so that we can make `ReflectResource` take `Into<FilteredResource>` and use this with reflection.  

It should be possible to make some `&'static Access<ComponentId>` for "read-all-resources" and "write-all-resources", but that requires some changes to `Access` to make it work in a `const` context and that seemed too big for this PR.  

This also keeps it more similar to `FilteredEntityRef`.

Reply:
> It should be possible to make some `&'static Access<ComponentId>` for "read-all-resources" and "write-all-resources", but that requires some changes to `Access` to make it work in a `const` context and that seemed too big for this PR.

You could make a `OnceCell`/`Lazy` for it, as long as it is either empty or contains "read-all-resources" and/or "write-all-resources" bools (i.e. no `ComponentId`s)

> This also keeps it more similar to `FilteredEntityRef`.

Unfortunately this kind of deisign is also creating problems for `FilteredEntityRef` due to the excessive cloning.

Reply:
Oh, I didn't think of using `OnceCell` or `Lazy`!  That's a good idea; I'll fix that.